### PR TITLE
reporting SMS delivery result for SendSmsInBackground method for Android via an event.

### PR DESCRIPTION
**Pull request by VahidShir**

### CODE REVIEW

File: Messaging/Plugin.Messaging.Abstractions/ISmsTask.cs [hunk 1]
```diff
@@ -45,5 +55,7 @@ public interface ISmsTask
         /// </remarks>
         void SendSmsInBackground(string recipient, string message = null);
 
+
+        event SmsDeliveryResult OnSmsDeliveryResult;
```
Comment (soroshsabz) [inline Messaging/Plugin.Messaging.Abstractions/ISmsTask.cs:58]: Please make comment and description on it

File: Messaging/Plugin.Messaging.Abstractions/ISmsTask.cs [hunk 2]
```diff
@@ -37,6 +46,7 @@ public interface ISmsTask
         /// </summary>
         /// <param name="recipient">Sms recipient</param>
         /// <param name="message">Sms message</param>
+        /// <param name="cancellationToken"></param>
```
Comment (soroshsabz) [inline Messaging/Plugin.Messaging.Abstractions/ISmsTask.cs]: This parameter does not exist in input params

File: Messaging/Plugin.Messaging.Abstractions/ISmsTask.cs [hunk 3]
```diff
@@ -1,5 +1,14 @@
...
+﻿using System.Threading;
+using System.Threading.Tasks;
+
+namespace Plugin.Messaging
 {
+    /// <summary>
+    /// 
+    /// </summary>
+    /// <param name="result"></param>
+    public delegate void SmsDeliveryResult(bool isSuccessful);
```
Comment (soroshsabz) [inline Messaging/Plugin.Messaging.Abstractions/ISmsTask.cs:7]: Why declare SmsDeliveryResult in out of ISmsTask?

File: Messaging/Plugin.Messaging.Android/SmsTask.cs [hunk 4]
```diff
@@ -6,15 +7,37 @@ namespace Plugin.Messaging
 {
     // NOTE: http://developer.xamarin.com/recipes/android/networking/sms/send_an_sms/
 
-    internal class SmsTask : ISmsTask
+    internal class SmsTask : BroadcastReceiver, ISmsTask
```
Comment (soroshsabz) [inline Messaging/Plugin.Messaging.Android/SmsTask.cs:11]: Did you can seperate BroadcastRecerver from SmsTask?

File: Messaging/Plugin.Messaging.Android/SmsTask.cs [hunk 5]
```diff
@@ -38,16 +61,57 @@ public void SendSms(string recipient = null, string message = null)
...
+            _smsManager = SmsManager.Default;
 
-            if (CanSendSmsInBackground)
+            var piSent = PendingIntent.GetBroadcast(CONTEXT, 0, new Intent(SMS_SENT_INTENT_ACTION), 0);
+            var piDelivered = PendingIntent.GetBroadcast(CONTEXT, 0, new Intent(SMS_DELIVERY_INTENT_ACTION), 0);
+
+            CONTEXT.RegisterReceiver(this, new IntentFilter(SMS_SENT_INTENT_ACTION));
+            CONTEXT.RegisterReceiver(this, new IntentFilter(SMS_DELIVERY_INTENT_ACTION));
+
+            _smsManager.SendTextMessage(recipient, null, message, piSent, piDelivered);
```
Comment (soroshsabz) [inline Messaging/Plugin.Messaging.Android/SmsTask.cs]: Why did you change send method? in prevoius we use SendMultipartTextMessage

**Review by soroshsabz — CHANGES_REQUESTED** [review]

File: Messaging/Plugin.Messaging.Abstractions/ISmsTask.cs (hunk 3 above)
Comment (VahidShir) [reply Messaging/Plugin.Messaging.Abstractions/ISmsTask.cs:7]: what's the problem of declaring it outside? Are there any specific concerns?
https://stackoverflow.com/a/5288262/5941852

File: Messaging/Plugin.Messaging.Android/SmsTask.cs (hunk 4 above)
Comment (VahidShir) [reply Messaging/Plugin.Messaging.Android/SmsTask.cs:11]: Separating SmsTask class from BroadcastRecerver class? Yeah I wanted tom but it was complex to me, and i didn't come up with an idea to do so. This was much more simpler to implement.

File: Messaging/Plugin.Messaging.Android/SmsTask.cs (hunk 5 above)
Comment (VahidShir) [reply Messaging/Plugin.Messaging.Android/SmsTask.cs]: **TD;DL:**
For Resa project this this new method is fine and is OK. But for other people we need to use previous method. But it would be a bit complex and time-consuming to implement that.

Because the result mechanism works only for one SMS. The previous method i.e SendMultipartTextMessage will accept message bodies in various lengths (even longer than one SMS limit) and then split them to be fit in one SMS limit. But in that case it would be more co,plex to me to support sending result(s) of multiple SMS. Because every single SMS has its own BroadcastReceiver event and combining all receiver events into one single boolean result would take more time to implement and it is a bit complex. The new send method is perfectly OK for RESA project, but for other people using this library I certainly need to use previous method and support getting result for multiple SMSs.

Comment (soroshsabz) [reply Messaging/Plugin.Messaging.Android/SmsTask.cs]: Please bring back.

File: Messaging/Plugin.Messaging.Android/SmsTask.cs [hunk 6]
```diff
@@ -38,16 +61,57 @@ public void SendSms(string recipient = null, string message = null)
...
-                var msgs = smsManager.DivideMessage(message);
-                smsManager.SendMultipartTextMessage(recipient, null, msgs, null, null);
             }
         }
 
         #endregion
+
+        #region Private Field
+
+        private const string SMS_DELIVERY_INTENT_ACTION = "BSN.Resa.DoctorApp.SMS_Delivery";
```
Comment (soroshsabz) [inline Messaging/Plugin.Messaging.Android/SmsTask.cs]: Why this specific project dependent string declared here? please remove it. or if you need it, please generals it

File: Messaging/Plugin.Messaging.Android/SmsTask.cs (hunk 4 above)
Comment (soroshsabz) [reply Messaging/Plugin.Messaging.Android/SmsTask.cs:11]: Please add another task in this project and link them to it, for future.

**Review by soroshsabz — CHANGES_REQUESTED** [review]

**Comment by soroshsabz** [comment]
@VahidShir any update?

File: Messaging/Plugin.Messaging.Android/SmsTask.cs (hunk 5 above)
Comment (VahidShir) [reply Messaging/Plugin.Messaging.Android/SmsTask.cs]: I did bring it back.
